// @progress/kendo-react-scheduler/dist/npm/context/SchedulerViewContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var utils_1 = require("../utils");
/** @hidden */
exports.SchedulerViewPropsContext = React.createContext({});
/** @hidden */
exports.useSchedulerViewPropsContext = function () { return React.useContext(exports.SchedulerViewPropsContext); };
exports.SchedulerViewPropsContext.displayName = 'SchedulerViewPropsContext';
/** @hidden */
exports.SchedulerViewRangesContext = React.createContext([]);
/** @hidden */
exports.useSchedulerViewRangesContext = function () { return React.useContext(exports.SchedulerViewRangesContext); };
exports.SchedulerViewRangesContext.displayName = 'SchedulerViewRangesContext';
/** @hidden */
exports.SchedulerViewSlotsContext = React.createContext([[], utils_1.noop]);
/** @hidden */
exports.useSchedulerViewSlotsContext = function () { return React.useContext(exports.SchedulerViewSlotsContext); };
exports.SchedulerViewSlotsContext.displayName = 'SchedulerViewSlotsContext';
/** @hidden */
exports.SchedulerViewItemsContext = React.createContext([[], utils_1.noop]);
/** @hidden */
exports.useSchedulerViewItemsContext = function () { return React.useContext(exports.SchedulerViewItemsContext); };
exports.SchedulerViewItemsContext.displayName = 'SchedulerViewItemsContext';
/** @hidden */
exports.SchedulerViewSelectedSlotsContext = React.createContext([[], utils_1.noop]);
/** @hidden */
exports.useSchedulerViewSelectedSlotsContext = function () { return React.useContext(exports.SchedulerViewSelectedSlotsContext); };
exports.SchedulerViewSelectedSlotsContext.displayName = 'SchedulerViewSelectedSlotsContext';
/** @hidden */
exports.SchedulerViewSelectedItemsContext = React.createContext([[], utils_1.noop]);
/** @hidden */
exports.useSchedulerViewSelectedItemsContext = function () { return React.useContext(exports.SchedulerViewSelectedItemsContext); };
exports.SchedulerViewSelectedItemsContext.displayName = 'SchedulerViewSelectedItemsContext';
/** @hidden */
exports.SchedulerViewFocusedItemsContext = React.createContext([[], utils_1.noop]);
/** @hidden */
exports.useSchedulerViewFocusedItemsContext = function () { return React.useContext(exports.SchedulerViewFocusedItemsContext); };
exports.SchedulerViewFocusedItemsContext.displayName = 'SchedulerViewFocusedItemsContext';
/** @hidden */
exports.SchedulerViewFocusedSlotsContext = React.createContext([[], utils_1.noop]);
/** @hidden */
exports.useSchedulerViewFocusedSlotsContext = function () { return React.useContext(exports.SchedulerViewFocusedSlotsContext); };
exports.SchedulerViewFocusedSlotsContext.displayName = 'SchedulerViewFocusedSlotsContext';
/** @hidden */
exports.SchedulerViewContext = function (_a) {
    var children = _a.children, 
    // Static
    props = _a.props, ranges = _a.ranges, 
    // Ref
    slots = _a.slots, items = _a.items, 
    // State
    // State.Selection
    selectedSlots = _a.selectedSlots, selectedItems = _a.selectedItems, 
    // State.Focus
    focusedItems = _a.focusedItems, focusedSlots = _a.focusedSlots;
    return (React.createElement(exports.SchedulerViewPropsContext.Provider, { value: props },
        React.createElement(exports.SchedulerViewRangesContext.Provider, { value: ranges },
            React.createElement(exports.SchedulerViewSlotsContext.Provider, { value: slots },
                React.createElement(exports.SchedulerViewItemsContext.Provider, { value: items },
                    React.createElement(exports.SchedulerViewSelectedSlotsContext.Provider, { value: selectedSlots },
                        React.createElement(exports.SchedulerViewSelectedItemsContext.Provider, { value: selectedItems },
                            React.createElement(exports.SchedulerViewFocusedItemsContext.Provider, { value: focusedItems },
                                React.createElement(exports.SchedulerViewFocusedSlotsContext.Provider, { value: focusedSlots }, children)))))))));
};
