// @progress/kendo-react-scheduler/dist/npm/context/SchedulerEditTaskContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
/** @hidden */
exports.SchedulerEditTaskRemoveItemContext = React.createContext([null]);
/** @hidden */
exports.useSchedulerEditTaskRemoveItemContext = function () { return React.useContext(exports.SchedulerEditTaskRemoveItemContext); };
/** @hidden */
exports.SchedulerEditTaskShowRemoveDialogContext = React.createContext([false]);
/** @hidden */
exports.useSchedulerEditTaskShowRemoveDialogContext = function () { return React.useContext(exports.SchedulerEditTaskShowRemoveDialogContext); };
/** @hidden */
exports.SchedulerEditTaskShowOccurrenceDialogContext = React.createContext([false]);
/** @hidden */
exports.useSchedulerEditTaskShowOccurrenceDialogContext = function () { return React.useContext(exports.SchedulerEditTaskShowOccurrenceDialogContext); };
/** @hidden */
exports.SchedulerEditTaskContext = function (props) {
    return (React.createElement(exports.SchedulerEditTaskRemoveItemContext.Provider, { value: props.remove },
        React.createElement(exports.SchedulerEditTaskShowRemoveDialogContext.Provider, { value: props.showRemoveDialog },
            React.createElement(exports.SchedulerEditTaskShowOccurrenceDialogContext.Provider, { value: props.showOccurrenceDialog }, props.children))));
};
